### ISSUE TITLE
support completions with multiple cursors

### ISSUE THREAD

**Issue by partumamet** [kiteco/issue-tracker#428]
Suppose we want to write the following code

```
my_dictionary = {}
assert a in my_dictionary
assert bc in my_dictionary
assert qqq in my_dictionary
```

first write
```
my_dictionary = {}
a
bc
qqq
```
Then in VSCode do multi line edit, and write `assert` in from of all the  terms. Then press "End" to go to the end of the line, and type `in my_` and the autocomplete should kick in at this stage, then press "tab".

The result will be
```
my_dictionary = {}
assert a in my_dictionary
assert bc in myassert a in my_dictionary
assert qqq in myassert a in my_dictionary
```
As you can see only the first line was autocompleted correctly. The 2nd and 3rd line results in the entire of the first line copied over.

I'm running Ubuntu 20.04 and VSCode v1.50.0 and Kite version 2.20201001.0

**Comment by metalogical** [comment]
Autocomplete for multiple cursors is currently unsupported, and this is expected behavior.

I suppose we could look into disabling Kite autocomplete entirely if you have multiple cursors – would this be better?

**Comment by partumamet** [comment]
Why not add the new feature to support multiline edit instead? 

Do you want users to uninstall Kite because it lacks support for a minor feature? =P

**Comment by metalogical** [comment]
So there are two issues at play here, which I outline below. Let me know if you disagree with my assessment. The summary though is that to make the behavior *useful* is not a "minor" feature, but actually a major one.

### The VS Code plugin exhibits buggy behavior.

* When multiple cursors exist, completions are requested at only one of the multiple cursors.
* But then when a resulting completion is selected, the completion should be inserted at only that one cursor, instead of the multiline insertion behavior you're seeing.
* The other cursors should remain unchanged.

So the the valid final output should be 

```
my_dictionary = {}
assert a in my_dictionary‸
assert bc in my‸
assert qqq in my‸
```

We can fix this. But it seems like it wouldn't be particularly useful, since if you wanted that completion, you would have just used a single cursor. The ideal behavior would require a fairly complex change to the Kite engine (see the next section). So I think the simplest thing is to just disable completions for multiple cursors.

### The Kite engine does not support multi-cursor completions

You might desire a completion that results in the buffer looking like

```
my_dictionary = {}
assert a in my_dictionary
assert bc in my_dictionary
assert qqq in my_dictionary
```

Unfortunately this is incompatible with how the Kite engine currently returns completions (of the form "replace this selection in the editor with this text"). Fixing this is a much more complex change, and is low priority for us.

**Comment by partumamet** [comment]
perhaps it is my ignorance, but I don't understand how multi-cursor behaves differently to single cursor? 

The exact same value returned for a single cursor can just be replicated for all multiple cursors, is that not true?

**Comment by metalogical** [comment]
> The exact same value returned for a single cursor can just be replicated for all multiple cursors, is that not true?

This is true in your example, but is not always true, and attempts at generalizing to other scenarios can potentially cause other bugs or just result in completions that don't make sense. The underlying form of the Kite completion is "replace this range of text with this other range of text," so generalizing that to multiple cursors would require shifting the range based on the other cursor positions, and you might imagine that this could result in some nasty edge cases.

> Perhaps instead of replacing the text (in the whole line) in the edit, simply perform replace of the text that triggered the autocompletion to occur

I'm not sure what you mean by this. Is this different from my proposal where we only insert the completion for the first line, and not the second or third?

**Comment by partumamet** [comment]
>>  Perhaps instead of replacing the text (in the whole line) in the edit, simply perform replace of the text that triggered the autocompletion to occur

>I'm not sure what you mean by this. Is this different from my proposal where we only insert the completion for the first line, and not the second or third?

Perhaps my understanding is not correct. I'm guessing the reason for the current observed behaviour I reported is because the entire first line is returned by the vscode plugin. Hence when applied to 2nd and 3rd line, we effectively get the entire first line repeated.

So is possible to not return the entire first line. Instead just the single "word" to be replaced, so the result would be correctly applied to the 2nd and 3rd line.

Eg. instead of the plugin returning "assert a in my_dictionary" and replacing the entire first line, it will simply return "dictionary" and insert it after "my_" instead.

**Comment by metalogical** [comment]
@partumamet this is possible, but undesirable right now, as since the Kite engine doesn't take multiple cursor semantics into account when generating completions, it would result in undesirable effects in *other* multiple cursor situations.

e.g. if we did this for the following buffer, `my_dictionary` would be suggested, but if we followed your insertion behavior, inserting the completion as you suggest would actually be incorrect.

```
my_dictionary = {}
assert a in my_‸
assert bc in ‸
assert qqq in ‸
```

**Comment by partumamet** [comment]
I tested this on SublimeText and looks like indeed Kite does not support multi-line edit.

It is a key important feature in my humble opinion. Really hoping that Kite will become much more powerful with nice usability features.

Unfortunately I'm not well versed with Plugins and JS. Otherwise, I would have forked the repo and tried a few things.
